src/core/server/services/users/users.ts

```diff
@@ -658,21 +659,110 @@ export async function updateUsernameByID(
 export async function updateRole(
   mongo: Db,
   tenant: Tenant,
-  user: Pick<User, "id">,
+  viewer: Pick<User, "id" | "role" | "moderationScopes">,
   userID: string,
   role: GQLUSER_ROLE
 ) {
-  if (user.id === userID) {
+  if (viewer.id === userID) {
     throw new Error("cannot update your own user role");
   }
 
+  // only admins can promote to admin
+  if (viewer.role !== GQLUSER_ROLE.ADMIN && role === GQLUSER_ROLE.ADMIN) {
+    throw new Error("only admins can update users to administrators");
+  }
+
+  // you can only update users if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(viewer.role)) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
```

Comment:
Not required and should be removed as it's currently duplicating logic from the schema.